HWP ACU checks for spin-up

jlashner opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
This PR adds a check of the ACU elevation before allowing for spinup of the HWP.

## Description
<!--- Describe your changes in detail -->
Adds ACU state information to the `HWPState` class. This state info is used by the control-state `update` function to check the ACU elevation and last update time before enabling spin-up.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Motivation is described in Kyohei's discussion here: https://github.com/simonsobs/daq-discussions/discussions/96.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
This has not yet been tested! @ykyohei can we test this on one of the SATs? To test at a safe elevation, we can raise the `acu_min_el` param.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

ykyohei commented:
Thank you so much for doing this quickly. Yes, we can test this.

I have a question. Do you plan to implement the constraints between gripper and ACU by implementing it directly in the gripper agent? Or do you plan to add gripper control in the supervisor function?

mhasself commented on socs/agents/hwp_supervisor/agent.py:
```diff
@@ -720,7 +776,20 @@ def query_pid_state():
+                    if acu.last_updated is None:
+                        raise RuntimeError(f"No ACU data has been received from instance-id {acu.instance_id}")
+                    tdiff = time.time() - acu.last_updated
+                    if tdiff < acu.max_time_since_update:
```
Sign error

mhasself commented on socs/agents/hwp_supervisor/agent.py:
```diff
@@ -182,6 +236,8 @@ def from_args(cls, args: argparse.Namespace):
         if args.driver_iboot_id is not None:
             self.driver_iboot = IBootState(args.driver_iboot_id, args.driver_iboot_outlets)
+        if args.acu_instance_id is not None:
+            self.acu = ACUState(args.acu_instance_id, args.acu_min_el, args.acu_max_time_since_update)
```
Needs acu_max_el, right?

Also, isn't it generally wise to use kwargs?

mhasself commented on socs/agents/hwp_supervisor/agent.py:
```diff
@@ -1457,6 +1528,23 @@ def make_parser(parser=None):
         '--gripper-iboot-outlets', nargs='+', type=int,
         help="Outlets for gripper iboot power")
 
+    pgroup.add_argument(
+        '--acu-instance-id', help="Instance ID for the ACU agent"
+    )
```
Somewhere, maybe many places, you need to mention that passing this is necessary for the ACU check to be active.  I.e. it's not sufficient to specify min-el and max-el.

ykyohei commented on socs/agents/hwp_supervisor/agent.py:
```diff
@@ -138,6 +138,58 @@ def update(self):
+@dataclass
+class ACUState:
+    """
+    Class containg ACU state information.
```
containing

mhasself commented on socs/agents/hwp_supervisor/agent.py:
```diff
@@ -182,6 +236,8 @@ def from_args(cls, args: argparse.Namespace):
             self.gripper_iboot = IBootState(args.gripper_iboot_id, args.gripper_iboot_outlets)
         if args.driver_iboot_id is not None:
             self.driver_iboot = IBootState(args.driver_iboot_id, args.driver_iboot_outlets)
+        if args.acu_instance_id is not None:
```
I would want to see log messages for each of these (not just ACU):
"Checking ACU elevation range is [NOT active / active at instance_id={instance_id}]"
etc.

When something goes wrong, whether the agent though these things were on, or not, is very useful log output.

ykyohei reviewed (commented):
`el_commanded_position` is monitored but not used for `check_acu_ok_for_spinup`. 
I think we should check this too.

jlashner commented:
@ykyohei My intention is to implement grip / ungrip procedures (or at least a wrapper for the gripper agent ops) in the supervisor, and to add the acu checks here. That will be the next PR.

ykyohei commented on socs/agents/hwp_supervisor/agent.py:
```diff
@@ -171,17 +223,41 @@ class HWPState:
             self.gripper_iboot = IBootState(args.gripper_iboot_id, args.gripper_iboot_outlets)
+            log.info("Gripper Ibootbar id set: {id}", id=args.gripper_iboot_id)
+        else:
+            log.warning("Gripper Ibootbar id not set")
```
log.warn not log.warning, same typo in several lines later

mhasself commented on socs/agents/hwp_supervisor/agent.py:
```diff
@@ -138,6 +138,58 @@ def update(self):
+        self.el_current_position = d['Elevation current position']
+        self.el_commanded_position = d['Elevation commanded position']
+        self.el_current_velocity = d['Elevation current velocity']
+        self.last_updated = time.time()
```
The timestamp should come from the ACU dataset.  Or at least be rejected if doesn't make sense.

I acknowledge the ACU session data isn't making that very easy... can you change this to
```
last_updated = d.get('timestamp_agent')
```
and we'll get that field added into the agent session.data separately?

jlashner replied:
Yep! Will `timestamp_agent` sometimes not exist / should I handle None values by just not changing `last_updated`?

mhasself replied:
Right, sorry, I meant that you could then replace None with time.time().  That's a compromise until the ACU agent is updated.  Once ACU agent is updated, I would expect timestamp_agent to be either absent (during startup) or a float.

mhasself reviewed (changes requested):
Just one more inline comment from me ... 

And my outline comment is that you should probably add some discussion of this feature to the docs, or else make the docstrings and argument descriptions clearer.  I can understand "Minimum elevation allowed before restricting spin-up" because I already know what this code is for... but I think there are clearer ways to say this.  ("Elevation below which HWP spin-up is not permitted", for example)

ykyohei commented:
I tested the acu_check in daq-dev and made small debugs. I believe it's working now.

jlashner commented:
Great, thanks Kyohei!

I just implemented Matthew's suggestion for acquiring the last_updated timestamp from ACU session data, and also added a brief section to the docs on ACU safety checks that we can expand as we add the gripper operations and checks.

mhasself reviewed (approved):
Thanks for the additional clarifications.

BrianJKoopman reviewed (commented):
This seems ready to merge, but there are some merge conflicts in the agent code. Can you resolve these @jlashner?

jlashner commented:
Looks like tests are failing due to this issue: https://github.com/simonsobs/ocs/issues/391

Do we need to pin setup-tools in socs or upgrade the ocs version?

BrianJKoopman commented:
> Looks like tests are failing due to this issue: [simonsobs/ocs#391](https://github.com/simonsobs/ocs/issues/391)
> 
> Do we need to pin setup-tools in socs or upgrade the ocs version?

I was thinking the later, but neither option is working...I might end up switching build systems, was trying to make that decision yesterday.
